Src/Base/AMReX_MultiFabUtil_1D_C.H

```diff
@@ -207,30 +208,32 @@ void amrex_avgdown (Box const& bx, Array4<Real> const& crse,
     }
 }
 
+template <typename T>
 AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-void amrex_avgdown (int i, int, int, int n, Array4<Real> const& crse,
-                    Array4<Real const> const& fine,
+void amrex_avgdown (int i, int, int, int n, Array4<T> const& crse,
+                    Array4<T const> const& fine,
                     int ccomp, int fcomp, IntVect const& ratio) noexcept
 {
     const int facx = ratio[0];
     const Real volfrac = 1.0/static_cast<Real>(facx);
```

Comment:
When `T` is float and `Real` is double, this will result in warning on float conversion later.  So maybe what you could do for Matrix3d is to specialize `amrex_avgdown` for Matrix3d.